compiler-user-guide/src/developing-hardware/limitations.md

```diff
@@ -0,0 +1,185 @@
+    Also, only certain types of recursion are considered non-synthesizable; recursively defined values are for example synthesizable: they are (often) synthesized to feedback loops.
+
+    Let us distinguish between three variants of recursion:
+
+    * __Dynamic data-dependent recursion__
+
+        As demonstrated in this definition of a function that calculates the n'th Fibbonacci number:
+
+        ``` haskell
+        fibR 0 = 0
+        fibR 1 = 1
+        fibR n = fibR (n-1) + fibR (n-2)
+        ```
+
+        To get the first 10 numbers, we do the following:
+
+        ```
+        >>> import qualified Data.List as L
+        >>> L.map fibR [0..9]
+        [0,1,1,2,3,5,8,13,21,34]
+        ```
+
+        The `fibR` function is not synthesizable by the Clash compiler, because, when we take a _structural_ view, `fibR` describes an infinitely deep structure.
+
+        In principal, descriptions like the above could be synthesized to a circuit, but it would have to be a _sequential_ circuit.
+        Where the most general synthesis would then require a stack.
+        Such a synthesis approach is also known as _behavioral_ synthesis, something which the Clash compiler simply does not do.
+        One reason that Clash does not do this is because it does not fit the paradigm that only functions working on values of type `Signal` result in sequential circuits, and all other (non higher-order) functions result in combinational circuits.
+        This paradigm gives the designer the most straightforward mapping from the original Haskell description to generated circuit, and thus the greatest control over the eventual size of the circuit and longest propagation delay.
+
+    * __Value-recursion__
+
+        As demonstrated in this definition of a function that calculates the n'th Fibbonaci number on the n'th clock cycle:
+
+        ``` haskell
+        fibS :: SystemClockResetEnable => Signal System (Unsigned 64)
+        fibS = r
+            where r = register 0 r + register 0 (register 1 r)
+        ```
+
+        To get the first 10 numbers, we do the following:
+
+        ```
+        >>> sampleN @System 11 fibS
+        [0,0,1,1,2,3,5,8,13,21,34]
+        ```
+
+        Unlike the `fibR` function, the above `fibS` function _is_ synthesizable by the Clash compiler.
+        Where the recursively defined (non-function) value _r_ is synthesized to a feedback loop containing three registers and one adder.
+
+        Note that not all recursively defined values result in a feedback loop.
+        An example that uses recursively defined values which does not result in a feedback loop is the following function that performs one iteration of bubble sort:
+
+        ``` haskell
+        sortVL xs = map fst sorted :< (snd (last sorted))
+          where
+            lefts  = head xs :> map snd (init sorted)
+            rights = tail xs
+            sorted = zipWith compareSwapL (lazyV lefts) rights
```

Comment:
There's an issue with this code: it is incomplete. It builds on an earlier example that is part of [Troubleshooting](https://hackage-content.haskell.org/package/clash-prelude-1.8.2/docs/Clash-Tutorial.html#g:18), and as such it comes out of nowhere here and omits the definition of `compareSwapL`.